### PR TITLE
[PD-1837] single page app - roles apis

### PR DESCRIPTION
**Pull request by lemniscate613**
99% of the work in this PR is in myjobs/views.py at 614 and below: https://github.com/DirectEmployers/MyJobs/blob/pd-1837-single-page-app-roles-apis/myjobs/views.py#L614

Please ignore any front-end work associated with /manage-users/. That work will be completed in another PR.

This PR has 6 API endpoints that are relevant for: https://directemployersfoundation.atlassian.net/browse/PD-1837

One is for listing activities:
* manage-users/api/activities/

And five are for CRUDing of roles:
* manage-users/api/roles/create/
* manage-users/api/roles/
* manage-users/api/roles/NUMBER
* manage-users/api/roles/edit/NUMBER
* manage-users/api/roles/delete/NUMBER

To do:
* create automated tests for these endpoints?
* connect these endpoints up into the frontend

[BTW, this Chrome extension is helpful for working with JSON: https://chrome.google.com/webstore/detail/jsonview/chklaanhfefbnpoihckbnefhakgolnmc?hl=en ]

To test each endpoint:

1) manage-users/api/activities/

Visit http://127.0.0.1:8000/manage-users/api/activities/

2) manage-users/api/roles/create/

Vist MyJobs local install: http://127.0.0.1:8000/
Open Chrome Dev Tools > Console

Copy/paste:

```javascript
// jQuery used for this example only
var jq = document.createElement('script');
jq.src = "//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js";
document.getElementsByTagName('head')[0].appendChild(jq);
var csrf = read_cookie("csrftoken");
```
And then:
```javascript
activities_array = ["delete user", "update user"]
users_array = ["dpoynter@apps.directemployers.org"]

$.post( "http://127.0.0.1:8000/manage-users/api/roles/create/",
  {
    csrfmiddlewaretoken: csrf, 
    role_name: "New Role",
    assigned_activities: activities_array,
    assigned_users: users_array
  },
  function( data ) {
    console.log( data );
});
```

Check that new role is created in myjobs_role table.

See that User has been associated with new role:
```python
User.objects.get(email="dpoynter@apps.directemployers.org").roles.all()
```
3) manage-users/api/roles/
Visit: http://127.0.0.1:8000/manage-users/api/roles/

4) manage-users/api/roles/NUMBER
Visit: http://127.0.0.1:8000/manage-users/api/roles/NUMBER

5) manage-users/api/roles/edit/NUMBER

Vist MyJobs local install: http://127.0.0.1:8000/

Copy/paste the following in Chrome Dev Tools > Console:
```javascript
// jQuery used for this example only
var jq = document.createElement('script');
jq.src = "//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js";
document.getElementsByTagName('head')[0].appendChild(jq);
var csrf = read_cookie("csrftoken");
```
And then:
```javascript
activities_array = ["delete user", "update user"]
users_array = ["dpoynter@apps.directemployers.org"]

$.post( "http://127.0.0.1:8000/manage-users/api/roles/edit/8/",
  {
    csrfmiddlewaretoken: csrf, 
    role_name: "New Role",
    assigned_activities: activities_array,
    assigned_users: users_array
  },
  function( data ) {
    console.log( data );
});
```
Then confirm role name changed with something like:
```python
Role.objects.get(pk=8)
```
Then confirm activities altered with something like:
```python
Role.objects.get(pk=8).activities.all()
```
6) manage-users/api/roles/delete/NUMBER

Delete one of your new roles:

```javascript
// jQuery used for this example only
var jq = document.createElement('script');
jq.src = "//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js";
document.getElementsByTagName('head')[0].appendChild(jq);
var csrf = read_cookie("csrftoken");
```
Then:
```javascript
$.ajax( "http://127.0.0.1:8000/manage-users/api/roles/delete/13/",
  {
    type: "DELETE",
    beforeSend: function(xhr) {
        xhr.setRequestHeader("X-CSRFToken", csrf);
    },
 success: function(data) { 
    console.log( data );
}});
```

### CODE REVIEW

File: templates/includes/topbar.html [hunk 1]
```diff
@@ -1,500 +0,0 @@
-{% load i18n %}
```
Comment (lemniscate613) [inline templates/includes/topbar.html]: Didn't mean to delete this file... I added it back locally.

**Comment by aspidites** [comment]
> manage-users/api/roles/create/
> manage-users/api/roles/
> manage-users/api/roles/NUMBER
> manage-users/api/roles/edit/NUMBER
> manage-users/api/roles/delete/NUMBER

I wonder if this would be a good opportunity to take advantage of using a RESTful API. So instead of the above, we'd have (in the same order):

POST manage-users/role
GET manage-users/roles
GET manage-users/role/%number
PUT manage-users/role/%number
DELETE manage-users/role/%number

File: .gitignore [hunk 2]
```diff
@@ -16,6 +16,10 @@ dev_*
 CACHE
 static/redirect.css
 
+Gulp stuff
```
Comment (aspidites) [inline .gitignore]: is this supposed to be commented out?

File: gulp/package.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "de-build",
+  "version": "0.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
```
Comment (aspidites) [inline gulp/package.json]: `"author": "DirectEmployers Association"`

File: gulp/package.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  "license": "All Rights Reserved",
```
Comment (aspidites) [inline gulp/package.json]: License should be GPL. See @darrint's original react PR.

File: gulp/package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    "browserify": "^11.2.0",
+    "es6-promise": "^3.0.2",
+    "fetch-polyfill": "^0.8.1",
+    "gulp": "^3.9.0",
+    "gulp-babel": "^5.3.0",
+    "gulp-jasmine": "^2.1.0",
+    "gulp-sourcemaps": "^1.6.0",
+    "gulp-uglify": "^1.4.2",
+    "gulp-util": "^3.0.7",
+    "immutable": "^3.7.5",
```
Comment (aspidites) [inline gulp/package.json]: I thought we weren't using immutible? Also, is es6-promise really just a dev dependency?

File: gulp/src/manageusers/manageusers.js [hunk 6]
```diff
@@ -0,0 +1,1022 @@
...
+var AssociatedUsersList = React.createClass({
+  getInitialState: function() {
+    return {
+      associated_users_list: ''
+    };
+  },
+  componentDidMount: function() {
+    if (this.isMounted()) {
+      var associated_users_list = [];
+      console.log("this.props.users is:");
```
Comment (aspidites) [inline gulp/src/manageusers/manageusers.js]: Are debug statements meant to still be here?

**Comment by aspidites** [comment]
I'm inclined to wait until after your todo items are complete before I bother testing this, as doing so will allow me to forgo the manual manipulation of variables.

File: gulp/src/manageusers/manageusers.js [hunk 7]
```diff
@@ -0,0 +1,1022 @@
...
+// This is the entry point of the application. Bundling begins here.
+
+
+
+
+
+var AssociatedUsersList = React.createClass({
+  getInitialState: function() {
+    return {
+      associated_users_list: ''
```
Comment (aspidites) [inline gulp/src/manageusers/manageusers.js]: Why is associated_users_list a string here, but an array later?

**Comment by aspidites** [comment]
So, this code seems to be very much ES5, where as @darrint's work is ES6. I don't think we should be mixing the two, as there is a bit of cognitive overload that happens when remembering that the scoping rules between the two versions of JS are different. I don't care whether we downgrade @darrint's stuff to ES5 or upgrade your stuff to ES6, but I think it should be consistent.

File: myjobs/views.py [hunk 8]
```diff
@@ -592,3 +595,319 @@ def topbar(request):
...
+    return HttpResponse(serializers.serialize("json", activities, fields=('name', 'description')))
+
+
+def api_get_roles(request):
+    """
+    GET /roles/
+    Retrieves all roles for a company
+    """
+
+    company_name = get_company_or_404(request)
```
Comment (aspidites) [inline myjobs/views.py]: get_comany_or_404 doesn't return a company name. It returns a company instance.

File: myjobs/views.py [hunk 9]
```diff
@@ -592,3 +595,319 @@ def topbar(request):
...
+    company_object = Company.objects.filter(name=company_name)
+    company_id = company_object[0].id
```
Comment (aspidites) [inline myjobs/views.py]: Which means this can just be `company_id = company.pk`

File: myjobs/views.py [hunk 10]
```diff
@@ -592,3 +595,319 @@ def topbar(request):
...
+
+    response_data = {}
+
+    # /api/roles/
+    # Return information about all roles of this company
+    roles = Role.objects.filter(company=company_id)
```
Comment (aspidites) [inline myjobs/views.py]: Though, you don't even need to pass just an id here, you can pass the company instance directly.

**Comment by darrint** [comment]
@aspidites It looks like this work is based on an earlier version of my node infrastructure, which is why licenses are wrong, etc.

File: gulp/src/manageusers/manageusers.js [hunk 11]
```diff
@@ -0,0 +1,1022 @@
...
+
+
+var RolesList = React.createClass({
+  getInitialState: function() {
+    return {
+      table_rows: ''
+    };
+  },
+  componentDidMount: function() {
+    $.get(this.props.source, function(results) {
```
Comment (darrint) [inline gulp/src/manageusers/manageusers.js]: I really do _not_ want to see code like this in React components. In my unit tests PR I put this code in a separate API class. But I was also using Redux so my whole approach is different. 

If we write code your way, with React state, I'm still of the opinion that we need to keep business logic out of the React components and push it into plain old JavaScript classes. Since you are using React state I don't know off-hand the right way to communicate with separate plain old JavaScript classes. The Redux way is that the plain old JavaScript classes communicate state to the components, never the other way around.

File: myjobs/views.py (hunk 8 above)
Comment (lemniscate613) [inline myjobs/views.py]: Updated - thank you.

File: myjobs/views.py (hunk 9 above)
Comment (lemniscate613) [inline myjobs/views.py]: (I'll try to respond to each comment.) Updated.

File: myjobs/views.py (hunk 10 above)
Comment (lemniscate613) [inline myjobs/views.py]: Thanks -- updated in 4 or 5 others places.

File: gulp/src/manageusers/manageusers.js (hunk 11 above)
Comment (lemniscate613) [inline gulp/src/manageusers/manageusers.js]: Understood. Though we had a good conversation about this yesterday, I think I'm going to respond to any comments made in a PR in that same PR.

The data flow in this app is simple because I wanted to get a basic version out during sprint 22. Now that you've established an awesome precedent (ie Redux, unit tests) I can fold it into this project too.

File: gulp/package.json (hunk 3 above)
Comment (lemniscate613) [inline gulp/package.json]: @darrint is correct -- this is based on an older version of his node infrastructure. You'll see things merged soon.

File: gulp/src/manageusers/manageusers.js (hunk 6 above)
Comment (lemniscate613) [inline gulp/src/manageusers/manageusers.js]: We had a good conversation about this in person yesterday, but I'll try to respond to all PR comments in that PR.

You're absolutely right -- I don't mean to leave these guys in. This PR was meant to draw attention solely to the API work though ("Please ignore any front-end work associated with /manage-users/.").

That said, I've got a nice workflow setup locally based on that conversation with @darrint yesterday:

* Work locally with console statements (to my :heart:'s delight)
* Continuously build with grunt watch-no-strip to leave those statements in (the default gulp build / gulp watch strips them out with a node package called gulp-strip-debug)
* Remove them manually before pushing a PR
* Finally, the gulpfile Jenkins uses to make the production build _can_ use that node package _just in case_ -- ensuring we don't have any console statements in the final product. That's the main purpose of the package (not doing your cleanup work for you, which is what I thought initially): https://github.com/sindresorhus/gulp-strip-debug

> Useful for making sure you didn't leave any logging in production code.

File: gulp/src/manageusers/manageusers.js (hunk 7 above)
Comment (lemniscate613) [inline gulp/src/manageusers/manageusers.js]: Nice catch. Updated.
